### DISCUSSION TITLE
Ikarus mods support?

### DISCUSSION THREAD

**Discussion by Elgcahlxukuth** [Try/OpenGothic#396] [Q&A]
Hi!

What is your view on supporting mods that use Ikarus (and LeGo)?

Are you planning to support such mods? Do you think it is possible to just execute such mods, without adjusting them to be able to work with OpenGothic?

I could see in another topic you mentioned that you execute the game with Chronicles of Myrtana .ini file, so what exactly do you achieve by that? Are the mods already supported to some extent? I understand that OpenGothic parses Gothic .d scripts, so any changes made to the game via the scripts will automatically be applied. But since Ikarus framework hacks the Gothic .exe memory, is it possible to map all the _instructions_ to data/functions in OpenGothic, so that the mods work?

I don't fully understand how exactly does the Ikarus achieve its goals, so I can't judge the difficulty/feasibility of supporting Ikarus mods in any Zengine reimplementation (like OpenGothic). I would gladly hear about this from someone more experienced :)

Cheers,
Elg'

**Comment by Try** [comment]
Hi, @Elgcahlxukuth !

> I don't fully understand how exactly does the Ikarus achieve its goals

In original Zengine, there are many bugs and some of them can be leveraged. For instance `[idx]` in  script implemented as `*(obj_ptr + mem_offset +idx*sizeof(T))`, with no range checks. This alone is enough for game mod to scout whole process memory and do stuff.
`Ikarus` uses more elegant way of memory hacking, yet principle stays the same. Common name for this volubility is DMA(direct memory access).
Script can use `int`, as 32-bit integer to represent memory pointers, since game run in 32bit mode.
On top of this Gothic.exe(and Gothic2.exe) that is shipped is debuggable executable. This allows to replace C++ vtbl with high level of reliability, with no worry for `inline` or other complications that optimization can bring to binary code.
Also debug symbols are useful for authors to build a map of game memory and leverage DMA with relative ease.

This may look like very ugly and fragile hack. And it is. Even, if someone (THQ) will use original C++ code to build new executable, all of this will break.

On OpenGothic side it impossible to have 100% support, but we may have some subset of features to allow for some content to run.

1. Ikarus functions `_^`, `_@` (`operator*()` and `operator&()` in C++ terms) are intercepted and handled by engine, but only a few ZEngine classes are implemented.
As states in https://github.com/Try/OpenGothic/issues/231 - only 10 of 136 items are working atm.
2. Stuff like inline assembler and *.dll load - won't do. For security reasons I would like to make VM as fully closed sandbox

> Chronicles of Myrtana .ini file, so what exactly do you achieve by that?

Not playable atm, using it for performance-profiling purposes and testing:
![изображение](https://user-images.githubusercontent.com/122894/215600087-81a71294-7956-41fb-8f99-8fc6d111b59f.png)
![изображение](https://user-images.githubusercontent.com/122894/215600112-72346c39-880d-4c59-bd2d-764caa54a2d3.png)
Not sure if noticeable, but sky and SSAO are quite broke, and shadows flickering, due to floating-point rounding.
Also ray-tracing crashes with device-lost (some bug in bindless code). 
So, nice testing for me :)
